### PR TITLE
[MLIR] XLS -> MLIR Translation

### PR DESCRIPTION
**Pull request by schilkp**
As requested, here is a combined PR that implements the XLS IR -> MLIR translation.

# STATE 

A few notes up front about the state of this:

- This relies on two PRs that are not yet merged - I have kept them in this PR branch so it is functional and will drop them + rebase once #1866 and #1843 are in.

- This currently implements full-fidelity round trip conversion for XLS functions. 

- Procs are mostly supported with the main limitation that non-zero initializers are not yet representable. This is WIP because it requires reworking of the yield OPs which is fairly invasive and I did not want commit to it until the "LiteralRegion" mechanism had been reviewed + merged.

- The notion of "top proc/func" is not yet translated/represented in MLIR.

- I have setup fuzzing of XLS->MLIR->XLS conversion that ties in interpretation of this round-tripped IR into the multi-way comparison:

  With this and the added unit tests I have a non-zero level of confidence that the translation for functions is correct :)

  Because proc initializers are not ready, I have not yet fuzzed those and expect there to be edge cases/bugs left.

  You can find the fuzzer for this branch here:

  https://github.com/schilkp/xls/tree/schilkp/xls_translate_func_fuzz

  To run it:
  ```bash
  bazel run -c opt //xls/fuzzer:run_fuzz_multiprocess -- --duration=500s --worker_count=32 --alsologtostderr --translate_ir_to_mlir=true --translate_opt_ir_to_mlir=true --translate_mlir_to_ir=true --crash_path=/tmp/fuzz_crasher --generate_proc=false
  ```

  I did not include the fuzzer in this commit because it still has some hacky workarounds. It touches core XLS infra so I assume it would be better to hold off there until it is ready.

  image

# WIDTH ATTRIBUTE

There is one MLIR Dialect design decision/bike-shedding discussion left related to this :)

There are now quite a few operations where the output width is not constraint
but relevant for the semantics of the operation. For example:

- Multiplication result can be any width, irregardless of operand types. The result is truncated or zero-padded if it does not fit. (This is also fixed in MLIR in this PR)
- Decode, where the result can be any size up less than or equal to `2**N` where `N` is the operand width.

XLS IR is not consistent in how operations with non-inferable return type are handle. For multiplication, the width of the output is simply taken from the result type:

```
literal.1: bits[8] = literal(value=7)
literal.2: bits[8] = literal(value=2)
ret result: bits[10] = umul(literal.1, literal.2)
```

For decode, the width is included as an "attribute":

```
onehot.10: bits[16] = decode(y, width=16, id=10)
```

Currently I updated the MLIR to behave exactly the same as XLS IR (required width for decode, no attr for multiplication etc).

In my eyes, it makes sense to unify this behavior at least in MLIR: Either all operations with unconstrained return type should have this information as an attribute, or all should simply use the actual return type.

Let me know what you think.

### CODE REVIEW

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 1]
```diff
@@ -0,0 +1,1708 @@
...
+    auto op = fn_map_.find(name);
+    if (op == fn_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown function name  ", name));
+    } else {
+      return op->second;
+    }
+  }
+
+  absl::Status setChn(std::string name, FlatSymbolRefAttr chn) {
+    if (chn_map_.contains(name)) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:82]: similarly apply the above two idioms to more functions please?

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 2]
```diff
@@ -0,0 +1,1708 @@
...
+      return absl::InternalError(absl::StrCat("Duplicate channel name ", name));
+    }
+    chn_map_[name] = chn;
+    return absl::OkStatus();
+  }
+
+  absl::StatusOr<SymbolRefAttr> getChn(std::string name) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:89]: Please call this getChannel ; avoid unnecessary abbreviations as they hinder readability.

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 3]
```diff
@@ -0,0 +1,1708 @@
...
+    auto op = chn_map_.find(name);
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:90]: similarly channel_map_

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 4]
```diff
@@ -0,0 +1,1708 @@
...
+  absl::StatusOr<Value> getMlirValue(int64_t id) {
+    auto op = ssa_map_.find(id);
+    if (op == ssa_map_.end()) {
+      return absl::InternalError(absl::StrCat("Unknown value id ", id));
+    } else {
+      return op->second;
+    }
+  }
+
+  void newFunctionBaseContext(bool is_proc) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:116]: no need to use this->

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 5]
```diff
@@ -0,0 +1,1708 @@
...
+    this->is_proc_ = is_proc;
+    this->state_elem_map = {};
+  }
+
+  void setStateElem(std::string name, Value elem) {
+    assert(is_proc_ && "Only proc functions contain state elements");
+    state_elem_map[name] = elem;
+  }
+
+  absl::StatusOr<Value> getStateElem(std::string id) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:127]: get/setStateElement?

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 6]
```diff
@@ -0,0 +1,1708 @@
...
+    } else {
+      return elem->second;
+    }
+  }
+
+ private:
+  // --- package/module state ----
+  // Maps name -> MLIR ref for all functions in current
+  // package/module.
+  std::unordered_map<std::string, FlatSymbolRefAttr> fn_map_;
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:141]: given you're using absl already, can you use absl::flat_hash_map instead of std::unordered_map everywhere please? absl::flat_hash_map has the same API as unordered_map but is significantly more performant.

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 7]
```diff
@@ -0,0 +1,1708 @@
...
+
+//===----------------------------------------------------------------------===//
+// Type Translation
+//===----------------------------------------------------------------------===//
+
+Type translateType(::xls::Type* xls_type, OpBuilder& builder,
+                   MLIRContext* ctx) {
+  switch (xls_type->kind()) {
+    case ::xls::TypeKind::kTuple: {
+      std::vector<Type> types{};
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:170]: {} not needed (any non-POD type has their default constructor called simply by `T t;`, dont need `T t{};`

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 8]
```diff
@@ -0,0 +1,1708 @@
...
+      for (auto type : xls_type->AsTupleOrDie()->element_types()) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:171]: generally when something returns a pointer, use auto* to make that explicit.

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 9]
```diff
@@ -0,0 +1,1708 @@
...
+                                builder.getIntegerAttr(type, converted_value))
+      .getOperation();
+}
+
+absl::StatusOr<Operation*> translateLiteral(const ::xls::Value& b,
+                                            OpBuilder& builder,
+                                            MLIRContext* ctx,
+                                            TranslationState& state) {
+  switch (b.kind()) {
+    case ::xls::ValueKind::kBits:
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:227]: I'd suggest adding `using ::xls::ValueKind;` at the top of this namespace to make these switches easier to read.

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 10]
```diff
@@ -0,0 +1,1708 @@
...
+}
+
+//===----------------------------------------------------------------------===//
+// Source Location Translation
+//===----------------------------------------------------------------------===//
+
+absl::StatusOr<Location> translateLoc(const ::xls::SourceInfo& xls_loc,
+                                      mlir::Builder& builder,
+                                      TranslationState& state) {
+  SmallVector<Location> locs{};
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:281]: again no need for {}

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 11]
```diff
@@ -0,0 +1,1708 @@
...
+    return opr_rhs.status();
+  }
+
+  auto result_type = translateType(node.GetType(), builder, ctx);
+  auto loc = translateLoc(node.loc(), builder, state);
+  if (!loc.ok()) {
+    return loc.status();
+  }
+
+  switch (node.op()) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:400]: There are sexier ways of doing this using TypeSwitch, but let's keep this as-is for now.

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 12]
```diff
@@ -0,0 +1,1708 @@
...
+  return builder.create<xls::CountedForOp>(
+      *loc, result_type, *initial_value, invar_args, trip_count, *body, stride);
+}
+
+absl::StatusOr<Operation*> translateAnyOp(::xls::Node& xls_node,
+                                          OpBuilder& builder, MLIRContext* ctx,
+                                          TranslationState& state) {
+  absl::StatusOr<Operation*> op;
+
+  if (auto xls_op = dynamic_cast<::xls::Literal*>(&xls_node)) {
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:1255]: You might take inspiration here from convertFunction in xls_translate.cc, the way it uses TypeSwitch. TypeSwitch dispatches to a lambda with an argument of the correct type, so then the compiler can pick the right overload.

e.g.

    return TypeSwitch<Operation*, BValue>(op)
        .Case<
            // Unary bitwise ops.
            IdentityOp, NotOp,
            // Variadic bitwise operations
            AndOp, OrOp, XorOp,
            // Arithmetic unary operations
            NegOp,
            // Binary ops.
            AddOp, SdivOp, SmodOp, SmulOp, SubOp, UdivOp, UmodOp, UmulOp,
            // Comparison operations
            EqOp, NeOp, SgeOp, SgtOp, SleOp, SltOp, UgeOp, UgtOp, UleOp, UltOp,
            // Shift operations
            ShllOp, ShrlOp, ShraOp,
            // Extension operations
            ZeroExtOp, SignExtOp,
            // Array ops.
            ArrayOp, ArrayZeroOp, ArrayIndexOp, ArrayIndexStaticOp,
            ArraySliceOp, ArrayUpdateOp, ArrayConcatOp,
            // Tuple operations
            TupleOp, TupleIndexOp,
            // Bit-vector operations
            BitSliceOp, BitSliceUpdateOp, DynamicBitSliceOp, ConcatOp,
            ReverseOp, DecodeOp, EncodeOp, OneHotOp,
            // Control-oriented operations
            SelOp, OneHotSelOp, PrioritySelOp,
            // Constants
            ConstantScalarOp, arith::ConstantOp,
            // Control flow
            mlir::func::CallOp, CountedForOp, MapOp,
            // Casts
            arith::BitcastOp, arith::IndexCastOp,
            // CSP ops
            AfterAllOp, SendOp, BlockingReceiveOp, NonblockingReceiveOp,
            // Debugging ops
            TraceOp>(
            [&](auto t) { return convertOp(t, translation_state, fb); })

**Review by jmolloy — CHANGES_REQUESTED** [review]

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc (hunk 11 above)
Comment (jmolloy) [reply xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:400]: Oh damn of course you're right. There's always std::visit, but frankly that might be grimmer (and is *awful* for compile time). Let's leave it.

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc (hunk 12 above)
Comment (jmolloy) [reply xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:1255]: Let's leave this, given conversation above.

**Comment by schilkp** [comment]
>     This relies on two PRs that are not yet merged - I have kept them in this PR branch so it is functional and will drop them + rebase once [[xls][mlir] Add LiteralOp to XLS MLIR Dialect. #1866](https://github.com/google/xls/pull/1866) and [[xls][mlir] Add missing MLIR dialect ops #1843](https://github.com/google/xls/pull/1843) are in.

They are now merged and this has been rebased.

**Review by jmolloy — APPROVED** [review]

File: xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc [hunk 13]
```diff
@@ -0,0 +1,1699 @@
...
+
+  // New function base scope
+  state.newFunctionBaseContext(/*is_proc=*/false);
+
+  // Add parameters to function context:
+  for (uint64_t arg_idx = 0; arg_idx < xls_func.params().length(); arg_idx++) {
+    auto xls_param = xls_func.params()[arg_idx];
+    auto mlir_arg = body->getArgument(arg_idx);
+    if (auto err = state.setMlirValue(xls_param->id(), mlir_arg); !err.ok()) {
+    }
```
Comment (jmolloy) [inline xls/contrib/mlir/tools/xls_translate/xls_translate_to_mlir.cc:1388]: if block is empty (LINT failure).

**Review by jmolloy — APPROVED** [review]
